pkg/kv/kvserver/replica_send.go

```diff
@@ -110,31 +112,18 @@ func (r *Replica) Send(
 // *StoreWriteBytes return value.
 func (r *Replica) SendWithWriteBytes(
 	ctx context.Context, ba roachpb.BatchRequest,
-) (*roachpb.BatchResponse, *StoreWriteBytes, *roachpb.Error) {
-	return r.sendWithoutRangeID(ctx, &ba)
+) (resp *roachpb.BatchResponse, swb *StoreWriteBytes, rErr *roachpb.Error) {
+	if r.store.cfg.Settings.CPUProfileType() != cluster.CPUProfileWithLabels {
+		return r.sendImpl(ctx, &ba)
+	}
+	pprof.Do(ctx, pprof.Labels("range_id", ba.RangeID.String()), func(ctx context.Context) {
```

Comment:
Should we add a comment with some rationale why this place is particularly good for adding labels?

Is there any place higher up the stack which could add this `RangeID` label instead? I.e. is `Replica.Send` the highest place where replica ID is defined?

Reply:
Replica.Send is the right place. We could move it one level higher (`Stores.Send`) but this wouldn't be as canonical IMO.

You could also write the code like this:

```go
if r.store.cfg.Settings.CPUProfileType() != cluster.CPUProfileWithLabels {
  prevCtx := ctx
  defer runtime.SetGoroutineLabels(prevCtx)
  ctx = pprof.WithLabels(ctx, pprof.Labels("range_str", r.rangeStr.String())
  runtime.SetGoroutineLabels(ctx)
}
return r.sendImpl(ctx, &ba)
```

I prefer this because it keeps it at a single caller for `sendImpl and makes sure the profiling doesn't complicate navigating call hierarchies.

Note that I also switched to `rangeStr` which is basically a concise print-out of the range descriptor, so it's more helpful than just the rangeID.

Reply:
Done, I agree that it's beneficial to do this way.